adapters/audienceNetwork/facebook.go

```diff
@@ -37,267 +41,416 @@ func (a *FacebookAdapter) Name() string {
+	imp := &out.Imp[0]
+	plmtId, pubId, err := this.extractPlmtAndPub(imp)
 	if err != nil {
-		err = &errortypes.BadServerResponse{
-			Message: err.Error(),
-		}
-		return
+		return err
 	}
-	if len(bidResp.SeatBid) == 0 {
-		return
+
+	reqExt := facebookReqExt{
+		PlatformId: this.platformId,
+		AuthId:     this.makeAuthId(out),
 	}
-	if len(bidResp.SeatBid[0].Bid) == 0 {
-		return
+
+	if out.Ext, err = json.Marshal(reqExt); err != nil {
+		return err
 	}
-	bid := bidResp.SeatBid[0].Bid[0]
 
-	result.Bid = &pbs.PBSBid{
-		AdUnitCode:        bid.ImpID,
-		Price:             bid.Price,
-		Adm:               bid.AdM,
-		CreativeMediaType: "banner", //  hard code this, because that's all facebook supports now, can potentially update it dynamically from "template" field in the "adm"
+	imp.TagID = pubId + "_" + plmtId
+	imp.Ext = nil
+
+	if out.App != nil {
+		app := *out.App
+		app.Publisher = &openrtb.Publisher{ID: pubId}
+		out.App = &app
+	} else {
+		site := *out.Site
+		site.Publisher = &openrtb.Publisher{ID: pubId}
+		out.Site = &site
 	}
-	return
+
+	if err = this.modifyImp(imp); err != nil {
+		return err
+	}
+
+	return nil
 }
 
-func (a *FacebookAdapter) MakeOpenRtbBidRequest(req *pbs.PBSRequest, bidder *pbs.PBSBidder, placementId string, mtype pbs.MediaType, pubId string, unitInd int) (openrtb.BidRequest, error) {
-	// this method creates imps for all ad units for the bidder with a single media type
-	fbReq, err := adapters.MakeOpenRTBGeneric(req, bidder, a.Name(), []pbs.MediaType{mtype})
+func (this *FacebookAdapter) modifyImp(out *openrtb.Imp) error {
+	impType, ok := resolveImpType(out)
+	if !ok {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("imp #%s with invalid type", out.ID),
+		}
+	}
 
-	if err != nil {
-		return openrtb.BidRequest{}, err
+	if out.Instl == 1 && impType != openrtb_ext.BidTypeBanner {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("imp #%s: interstitial imps are only supported for banner", out.ID),
+		}
 	}
 
-	fbReq.Ext = a.platformJSON
+	switch impType {
+	case openrtb_ext.BidTypeBanner:
+		if out.Instl == 1 {
+			out.Banner.W = openrtb.Uint64Ptr(0)
+			out.Banner.H = openrtb.Uint64Ptr(0)
+			out.Banner.Format = nil
+
+			return nil
+		}
+
+		if _, ok := supportedBannerHeights[*out.Banner.H]; !ok {
+			return &errortypes.BadInput{
+				Message: fmt.Sprintf("imp #%s: only banner heights 50 and 250 are supported", out.ID),
+			}
+		}
+
+		/* This will get overwritten post-serialization */
+		out.Banner.W = openrtb.Uint64Ptr(0)
+		out.Banner.Format = nil
+		break
 
-	if fbReq.Imp != nil && len(fbReq.Imp) > 0 {
-		// only returns 1 imp for requested ad unit
-		fbReq.Imp = fbReq.Imp[unitInd : unitInd+1]
+	default:
+		break
+	}
+
+	return nil
+}
 
-		if fbReq.Site != nil {
-			siteCopy := *fbReq.Site
-			siteCopy.Publisher = &openrtb.Publisher{ID: pubId}
-			fbReq.Site = &siteCopy
+func (this *FacebookAdapter) extractPlmtAndPub(out *openrtb.Imp) (string, string, error) {
```

Comment:
it took me a while to figure out what plmt was referring to. Consider renaming to `extractPlacementAndPub` to make it easier to read.

Reply:
done